README
= Reggie B. -- A Hex/RGB Color Converter

== Description

This is not the gem that you want.  You want the full-featured "color" Ruby gem.

You're still here?  Okay, this is what this gem is all about:

Reggie B. is a utility for converting colors between rgb and hex format.  It was created
to support http://www.2rgb.com.  It is especially useful for handling CSS color formats.

== Examples

  ReggieB.convert "rgb(255, 255, 255)"     =>  "0xffffff"
  ReggieB.convert "255, 255, 255"          =>  "0xffffff"
  ReggieB.convert "rgb(100%, 100%, 100%)"  =>  "0xffffff"
  ReggieB.convert "0xffffff"               =>  [255, 255, 255]
  ReggieB.convert "#ffffff"                =>  [255, 255, 255]
  ReggieB.convert "#fff"                   =>  [255, 255, 255]


CHANGELOG
0.1.0 (October 2, 2009)
* Added ability to accept rgb colors specified as percentages:
  rgb(50%, 34%, 12%)

0.0.1 (September 27, 2009)
* Initial release.


test/reggieb_test.rb
require 'test/test_helper'
require 'reggieb'

class ReggieBTest < Test::Unit::TestCase

  def test_simple_rgb_to_hex
    assert_equal "0xffffff", ReggieB.to_hex(255, 255, 255)
    assert_equal "0x000000", ReggieB.to_hex(0, 0, 0)
    assert_equal "0x00000f", ReggieB.to_hex(0, 0, 15)
    assert_equal "0x001100", ReggieB.to_hex(0, 17, 0)
    assert_equal "0xa01100", ReggieB.to_hex(160, 17, 0)
  end

  def test_simple_rgb_to_hex_errors
    assert_raise ArgumentError do
      ReggieB.to_hex(255, 256, 255)
    end
    assert_raise ArgumentError do
      ReggieB.to_hex(-10, 10, 10)
    end
    assert_raise ArgumentError do
      ReggieB.to_hex(8, 9, 1000)
    end
  end

  def test_simple_hex_to_rgb
    assert_equal  [255, 255, 255],  ReggieB.to_rgb("0xffffff")
    assert_equal  [0, 0, 0],        ReggieB.to_rgb("0x000000")
    assert_equal  [0, 0, 15],       ReggieB.to_rgb("0x00000f")
    assert_equal  [0, 17, 0],       ReggieB.to_rgb("0x001100")
    assert_equal  [160, 17, 0],     ReggieB.to_rgb("0xa01100")
    assert_equal  [0, 17, 0],       ReggieB.to_rgb("001100")
    assert_equal  [160, 17, 0],     ReggieB.to_rgb("a01100")
    assert_equal  [160, 17, 0],     ReggieB.to_rgb("A01100")
    assert_equal  [255, 255, 255],  ReggieB.to_rgb("0xFFFFFF")
  end

  def test_simple_hex_to_rgb_errors
    assert_raise ArgumentError do
      ReggieB.to_rgb("0xfff")
    end
    assert_raise ArgumentError do
      ReggieB.to_rgb("fff")
    end
    assert_raise ArgumentError do
      ReggieB.to_rgb("forks")
    end
    assert_raise ArgumentError do
      ReggieB.to_rgb("0x7897zq")
    end
  end
  
  def test_six_digit_hex
    assert_equal  [255, 255, 255],  ReggieB.convert("0xffffff")
    assert_equal  [0, 0, 0],        ReggieB.convert("0x000000")
    assert_equal  [0, 0, 15],       ReggieB.convert("0x00000f")
    assert_equal  [0, 17, 0],       ReggieB.convert("0x001100")
    assert_equal  [160, 17, 0],     ReggieB.convert("0xa01100")
    assert_equal  [0, 17, 0],       ReggieB.convert("001100")
    assert_equal  [160, 17, 0],     ReggieB.convert("a01100")
    assert_equal  [160, 17, 0],     ReggieB.convert("A01100")
    assert_equal  [255, 255, 255],  ReggieB.convert("0xFFFFFF")
  end

  def test_three_digit_hex
    assert_equal  [255, 255, 255],  ReggieB.convert("0xfff")
    assert_equal  [0, 0, 0],        ReggieB.convert("0x000")
    assert_equal  [0, 0, 255],      ReggieB.convert("0x00f")
    assert_equal  [0, 17, 0],       ReggieB.convert("0x010")
    assert_equal  [34, 0, 170],     ReggieB.convert("0x20a")
    assert_equal  [255, 255, 255],  ReggieB.convert("0xFFF")
    assert_equal  [0, 0, 255],      ReggieB.convert("0x00F")
    assert_equal  [34, 0, 170],     ReggieB.convert("0x20A")
    assert_equal  [255, 255, 255],  ReggieB.convert("FFF")
    assert_equal  [0, 0, 255],      ReggieB.convert("00F")
    assert_equal  [34, 0, 170],     ReggieB.convert("20A")
    assert_equal  [255, 255, 255],  ReggieB.convert("fff")
  end

  def test_wrong_digit_hex
    assert_raise ArgumentError do
      ReggieB.convert("0xffff")
    end
    assert_raise ArgumentError do
      ReggieB.convert("0x12345")
    end
    assert_raise ArgumentError do
      ReggieB.convert("0x123456789")
    end
    assert_raise ArgumentError do
      ReggieB.convert("0x1")
    end
    assert_raise ArgumentError do
      ReggieB.convert("0x")
    end
  end

  def test_hex_prefixes
    assert_equal  [255, 255, 255],  ReggieB.convert("0xffffff")
    assert_equal  [0, 170, 0],      ReggieB.convert("0x0a0")
    assert_equal  [0, 0, 11],       ReggieB.convert("0x00000B")
    assert_equal  [255, 255, 255],  ReggieB.convert("ffffff")
    assert_equal  [0, 170, 0],      ReggieB.convert("0a0")
    assert_equal  [0, 0, 11],       ReggieB.convert("00000B")
    assert_equal  [255, 255, 255],  ReggieB.convert("#ffffff")
    assert_equal  [0, 170, 0],      ReggieB.convert("#0a0")
    assert_equal  [0, 0, 11],       ReggieB.convert("#00000B")
  end

  def test_whitespace
    assert_equal  [255, 255, 255],  ReggieB.convert(" 0xffffff ")
    assert_equal  [0, 170, 0],      ReggieB.convert("    0x0a0    ")
    assert_equal  [0, 0, 11],       ReggieB.convert("0x00000B  ")
    assert_equal  [255, 255, 255],  ReggieB.convert(" ffffff")
    assert_equal  [0, 170, 0],      ReggieB.convert("\t0a0")
    assert_equal  [0, 0, 11],       ReggieB.convert("\n00000B\t")
    assert_equal  [255, 255, 255],  ReggieB.convert("               #ffffff")
    assert_equal  [0, 170, 0],      ReggieB.convert("#0a0\n")
    assert_equal  [0, 0, 11],       ReggieB.convert(" #00000B ")
  end

  def test_bad_hex_digits
    assert_raise ArgumentError do
      ReggieB.convert("0xbcq")
    end
    assert_raise ArgumentError do
      ReggieB.convert("ffg")
    end
    assert_raise ArgumentError do
      ReggieB.convert("0x0967uf")
    end
  end
  
  def test_semicolons
    assert_equal  [255, 255, 255],  ReggieB.convert("0xffffff;")
    assert_equal  [0, 170, 0],      ReggieB.convert("    0x0a0;")
    assert_equal  [0, 0, 11],       ReggieB.convert("0x00000B;  ")
    assert_equal  [255, 255, 255],  ReggieB.convert(" ffffff;")
    assert_equal  [0, 170, 0],      ReggieB.convert("\t0a0    ;")
    assert_equal  [0, 0, 11],       ReggieB.convert("\n00000B\t;")
    assert_equal  [255, 255, 255],  ReggieB.convert("               #ffffff;")
    assert_equal  [0, 170, 0],      ReggieB.convert("#0a0;")
    assert_equal  [0, 0, 11],       ReggieB.convert("#00000B;")
  end

  def test_simple_rgb
    assert_equal "0xffffff", ReggieB.convert("255, 255, 255")
    assert_equal "0x000000", ReggieB.convert("0, 0, 0")
    assert_equal "0x00000f", ReggieB.convert("0, 0, 15")
    assert_equal "0x001100", ReggieB.convert("0, 17, 0")
    assert_equal "0xa01100", ReggieB.convert("160, 17, 0")
  end

  def test_rgb_whitespace
    assert_equal "0xffffff", ReggieB.convert("255,255,255")
    assert_equal "0x000000", ReggieB.convert("       0, 0, 0        ")
    assert_equal "0x00000f", ReggieB.convert("0\t0\t15")
    assert_equal "0x001100", ReggieB.convert("\t0        17,0")
    assert_equal "0xa01100", ReggieB.convert("  160, 17, 0  ")
  end

  def test_rgb_semicolons
    assert_equal "0xffffff", ReggieB.convert("255,255,255;")
    assert_equal "0x000000", ReggieB.convert("       0, 0, 0;        ")
    assert_equal "0x00000f", ReggieB.convert("0\t0\t15;")
    assert_equal "0x001100", ReggieB.convert("\t0        17,0;")
    assert_equal "0xa01100", ReggieB.convert("  160, 17, 0  ;")
  end

  def test_rgb_delimeters
    assert_equal "0xffffff", ReggieB.convert("255-255-255")
    assert_equal "0x000000", ReggieB.convert("0 0 0")
    assert_equal "0x00000f", ReggieB.convert("0   0   15")
    assert_equal "0x001100", ReggieB.convert("rgb(0, 17, 0)")
    assert_equal "0xa01100", ReggieB.convert("rgb(160, 17, 0);")
  end

  def test_rgb_percentages
    assert_equal "0xffffff", ReggieB.convert("rgb(100%, 100%, 100%)")
    assert_equal "0xff7f7f", ReggieB.convert("rgb(100%, 50%, 50%)")
    assert_equal "0x000000", ReggieB.convert("rgb(0%, 0%, 0%)")
    assert_equal "0xffffff", ReggieB.convert("rgb(100%, 100%, 100%);")
    assert_equal "0x7fff7f", ReggieB.convert("50% 100% 50%")
    assert_equal "0x7fff7f", ReggieB.convert("    50%      100%     50%   ")
    assert_equal "0x7fff7f", ReggieB.convert("50%100%50%")
  end

end


lib/reggieb.rb
class ReggieB

  # converts rgb colors to hex colors and vice versa.  The input is a string representing
  # an rgb or hex color.  Many formats are allowed, including:
  #
  # - "rgb(255, 255, 255)"
  # - "rgb(255, 255, 255);"
  # - "rgb 255, 255, 255 "
  # - "(255, 255, 255)"
  # - "255, 255, 255"
  # - "255 255 255"
  # - "rgb(100%, 100%, 100%)"
  # - "100% 100% 100%"
  # - "0xffffff"
  # - "0xFFFFFF"
  # - "#ffffff"
  # - "ffffff"
  # - "fff"
  #
  # In the case only three hex digits are given, the three digits are assumed to expand
  # into six digits like so:
  #
  # - fff => ffffff
  # - 78e => 7788ee
  # - 084 => 008844
  # - ace => aaccee
  #
  # The output is either a hex color or an rgb color.  A hex color is represented as a 
  # string formatted like "0xdddddd" where d represents a hex digit.  An rgb color is
  # represented as an array of three integers between 0 and 255.
  #
  def self.convert(s)
    s = s.downcase
    s.gsub!(';', '')
    if s =~ /^\s*((0x|#|)[0-9,a-f]{6})\s*$/
      return to_rgb($1)
    end
    if s =~ /^\s*(0x|#|)([0-9,a-f])([0-9,a-f])([0-9,a-f])\s*$/
      s = $2 + $2 + $3 + $3 + $4 + $4
      return to_rgb(s)
    end
    if s =~ /^\D*(\d{1,3})%\D*(\d{1,3})%\D*(\d{1,3})%\D*$/
      percent_to_byte = 255.0 / 100.0
      return to_hex($1.to_i*percent_to_byte, $2.to_i*percent_to_byte, $3.to_i*percent_to_byte)
    end
    if s =~ /^\D*(\d{1,3})\D+(\d{1,3})\D+(\d{1,3})\D*$/
      return to_hex($1.to_i, $2.to_i, $3.to_i)
    end
    raise ArgumentError.new("Bad color format.")
  end

  # converts an rgb color to hex.  rgb values must be between 0 and 255.  The output
  # is a string formatted like "0xdddddd" where d represents a hex digit.
  #
  def self.to_hex(r, g, b)
    r = r.round
    g = g.round
    b = b.round
    unless (0..255).include?(r) && (0..255).include?(g) && (0..255).include?(b)
      message = "RGB values should be between 0 and 255.  Received #{r}, #{g}, #{b}."
      raise ArgumentError.new(message)
    end
    "0x#{to_hex_byte(r)}#{to_hex_byte(g)}#{to_hex_byte(b)}"
  end
  
  # converts a hex color to rgb.  The hex string should be formatted as "0xdddddd" or
  # "dddddd" where d represents a hex digit.  An array of three integers is returned
  # representing red, green, and blue values between 0 and 255.
  #
  def self.to_rgb(hex)
    hex = hex.downcase
    if hex !~ /0x[0-9,a-f]{6}/ && hex !~ /[0-9,a-f]{6}/
      raise ArgumentError.new("Hex number should be formatted as 0xdddddd or dddddd.")
    end
    last_byte   = hex[-2, 2]
    middle_byte = hex[-4, 2]
    first_byte  = hex[-6, 2]
    return [first_byte.to_i(16), middle_byte.to_i(16), last_byte.to_i(16)]
  end

  def self.to_hex_byte(digits)
    hex_string = digits.to_s(16)
    if hex_string.length == 1
      "0" + hex_string
    else
      hex_string
    end
  end
  private_class_method :to_hex_byte

end
